src/ray/rpc/grpc_callback_client.h

```diff
@@ -0,0 +1,336 @@
+            &SERVICE::Stub::experimental_async::METHOD),                                \
+        callback, ctx, #SERVICE ".grpc_client." #METHOD);                               \
+  }
+
+///
+/// The callback-based gRPC client.
+///
+
+template <class GrpcService>
+class GrpcCallbackClient {
+ public:
+  GrpcCallbackClient(const std::string &address, const int port,
+                     instrumented_io_context &io_context, int num_threads = 1)
+      : io_context_(io_context) {
+    grpc::ChannelArguments argument;
+    if (num_threads > 1) {
+      grpc::ResourceQuota quota;
+      quota.SetMaxThreads(num_threads);
+      argument.SetResourceQuota(quota);
+    }
+    // Disable http proxy since it disrupts local connections. TODO(ekl) we should make
+    // this configurable, or selectively set it for known local connections only.
+    argument.SetInt(GRPC_ARG_ENABLE_HTTP_PROXY, 0);
+    argument.SetMaxSendMessageSize(RayConfig::instance().max_grpc_message_size());
+    argument.SetMaxReceiveMessageSize(RayConfig::instance().max_grpc_message_size());
+    std::shared_ptr<grpc::Channel> channel =
+        grpc::CreateCustomChannel(address + ":" + std::to_string(port),
+                                  grpc::InsecureChannelCredentials(), argument);
+    stub_ = GrpcService::NewStub(channel);
+  }
+
+  /// Call a unary RPC method.
+  ///
+  /// \tparam Request Type of the request message.
+  /// \tparam Reply Type of the reply message.
+  ///
+  /// \param[in] method Pointer to the gRPC-generated `FooService::Stub::async::METHOD`
+  /// function.
+  /// \param[in] request The request message.
+  /// \param[in] callback The callback function that handles the reply.
+  /// \param[in] ctx The client context for this call. If not given, a
+  /// default-constructed context will be used.
+  /// \param[in] call_name A human-readable name for this call that will be used for
+  /// event loop statistics collection.
+  template <class Request, class Reply>
+  void CallUnaryMethod(const AsyncUnaryMethod<GrpcService, Request, Reply> method,
+                       const Request &request, const ClientCallback<Reply> &callback,
+                       std::shared_ptr<grpc::ClientContext> ctx = nullptr,
+                       const std::string call_name = "UNKNOWN_RPC") {
+    // We use shared pointers for the client context and response in order to let the
+    // below closure captures manage when this data is freed. This assumes that the
+    // client context and the response will no longer be used by the gRPC library after
+    // the gRPC-level callback is called (currently true) and is why we can provide gRPC
+    // raw pointers without breaking the shared pointers' reference counting.
+    if (ctx == nullptr) {
+      ctx = std::make_shared<grpc::ClientContext>();
+    }
+    auto response = std::make_shared<Reply>();
+    auto stats_handle = io_context_.RecordStart(call_name);
+    (stub_->experimental_async()->*method)(
```

Comment:
Looks like this experimental flag is gone now?

Reply:
Yep, once we update gRPC again we'll be able to use the callback/reactor API from the stable namespace.